src/Induction/NoInfiniteDescent.agda

```diff
@@ -0,0 +1,81 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- A standard consequence of accessibility/well-foundedness:
+-- the impossibility of 'infinite descent' from any (accessible)
+-- element x satisfying P to 'smaller' y also satisfying P
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Induction.NoInfiniteDescent where
+
+open import Data.Product.Base using (_,_; ∃-syntax; _×_)
+open import Function.Base using (_∘_)
+open import Induction.WellFounded using (WellFounded; Acc; acc-inverse; module Some)
+open import Level using (Level)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Nullary.Negation.Core using (¬_)
+open import Relation.Unary using (Pred; _∩_)
+
+private
+  variable
+    a r ℓ : Level
+    A : Set a
+
+------------------------------------------------------------------------
+-- Definitions
+
+module InfiniteDescent (_<_ : Rel A r) (P : Pred A ℓ)  where
+
+    DescentAt : Pred A _
+    DescentAt x = P x → ∃[ y ] y < x × P y
+
+    InfiniteDescent : Set _
+    InfiniteDescent = ∀ {x} → DescentAt x
```

Comment:
How is this 'infinite' ? Who tells me that `A` actually has infinitely many inhabitants?

It seems to me that this predicate is true basically because `DescentAt x` is always false if you pick `x` to be a 'top' element.

I expected some kind of map involving Nat that traced an infinitely long path.

Reply:
You make a very fair point. 

Originally, I had called the `InfiniteDescent` definition `SmallerCounterexample`, and the overall module `NoSmallestCounterexample`, because that's the to-me-more-familiar name for the principle. But Archimedes' "method of infinite descent" (from "The Sand Reckoner", IIRC), as (re)discovered by Fermat in the 17th century seems to be the more accepted term in (classical) mathematics (where of course `P` is a *negated* proposition, corresponding to "counterexamples to P", so the conclusion is `All P` from the impossibility of `¬ P`... but I digress).

In the 20th century, I suppose we might/should call this "Noetherian induction", and the property of being 'Noetherian' being also usually called "no infinite descending chains", because *in the presence of countable dependent choice*, one may iterate the `InfiniteDescent` condition precisely to generate a Nat-indexed descending chain in the ordering. But perhaps we *do* have enough from the existential to iterate the process? I'll try.